device: Update SYS_INIT_NAMED(), Z_DEVICE_INIT_ENTRY_DEFINE()  to be C++ compatible

tristan-google opened the pull request:
In #84394, `struct init_entry` was modified to remove an unneeded union. The `SYS_INIT_NAMED()` macro was adjusted accordingly, but is no longer C++ compatible due to the partial designated initializer.

Add an explicit value (NULL) for the other field (`dev`) in that struct.

Additional change (second commit): the `Z_DEVICE_INIT_ENTRY_DEFINE()` macro uses a designated initializer to define a struct. Ensure all members of `struct init_entry` are defined in order for C++ compatibility.

tristan-google committed:
device: Update SYS_INIT_NAMED() to be C++ compatible

In #84394, `struct init_entry` was modified to remove an unneeded union.
The `SYS_INIT_NAMED()` macro was adjusted accordingly, but is no longer
C++ compatible due to the partial designated initializer.

Add an explicit value (NULL) for the other field (`dev`) in that struct.

Signed-off-by: Tristan Honscheid <honscheid@google.com>

fabiobaltieri commented on include/zephyr/init.h:
```diff
@@ -163,10 +163,10 @@ struct init_entry {
-		Z_INIT_ENTRY_SECTION(level, prio, 0) __used __noasan                      \
-		Z_INIT_ENTRY_NAME(name) = {.init_fn = (init_fn_)}                         \
+#define SYS_INIT_NAMED(name, init_fn_, level, prio)                                                \
+	static const Z_DECL_ALIGN(struct init_entry) Z_INIT_ENTRY_SECTION(level, prio, 0)          \
+		__used __noasan                                                                    \
```
Does this need to reformat these two lines? Conventionally you'd avoid that.

fabiobaltieri commented:
Feels like we are going in circles with this one:
- 04611a5735ed51fe62c77216bceff7b8d5e3b9dc
- 153f38a412f87aa652a1857ab83df2f4f9746326
- 2438dbb613731f1abb90be9e1b4c7bd529d46574
- 19a33c7884134070ab9a13cf8112f5c8f81c5b94
- c15f029a7108117ab62d3275b77cddd8d77e0b81
- ed3377ae7815a3d1a73520c7e64f79f8b0a90d63

Shouldn't c++ tests catch this? At least in the push run.

gmarull reviewed (changes requested):
please remove stray formatting changes, otherwise lgtm

tristan-google committed:
device: Update Z_DEVICE_INIT_ENTRY_DEFINE to be C++ compatible

The `Z_DEVICE_INIT_ENTRY_DEFINE()` macro uses a designated initializer
to define a struct. Ensure all members of `struct init_entry` are
defined in order for C++ compatibility.

Signed-off-by: Tristan Honscheid <honscheid@google.com>
